source/hpack.c

```diff
@@ -255,48 +310,71 @@ void aws_hpack_context_destroy(struct aws_hpack_context *context) {
-    if (index == 0 || index >= s_static_header_table_size + context->dynamic_table.num_elements) {
+const struct aws_http_header *aws_hpack_get_header(struct aws_hpack_context *context, uint64_t index) {
+    if (index == 0 || index >= STATIC_HEADER_TABLE_SIZE + context->dynamic_table.num_elements) {
         aws_raise_error(AWS_ERROR_INVALID_INDEX);
         return NULL;
     }
 
     /* Check static table */
-    if (index < s_static_header_table_size) {
+    if (index < STATIC_HEADER_TABLE_SIZE) {
         return &s_static_header_table[index];
     }
 
     /* Check dynamic table */
-    index -= s_static_header_table_size;
+    index -= STATIC_HEADER_TABLE_SIZE;
     AWS_ASSERT(index < context->dynamic_table.num_elements);
     return &context->dynamic_table
                 .buffer[(context->dynamic_table.index_0 + index) % context->dynamic_table.max_elements];
 }
 
-int aws_hpack_find_index(struct aws_hpack_context *context, const struct aws_http_header *header, size_t *index) {
+int aws_hpack_find_index(
+    struct aws_hpack_context *context,
+    const struct aws_http_header *header,
+    uint64_t *index,
+    bool *found_value) {
+
+    *index = 0;
+    *found_value = false;
 
     /* Check static table */
     struct aws_hash_element *elem = NULL;
     aws_hash_table_find(&s_static_header_reverse_lookup, header, &elem);
     if (elem) {
-        *index = (size_t)elem->value;
+        *index = (uint64_t)elem->value;
+        *found_value = ((const struct aws_http_header *)elem->key)->value.len;
+        return AWS_OP_SUCCESS;
+    }
+    /* If not found, check name only table. Don't set found_value, it will be false */
+    aws_hash_table_find(&s_static_header_reverse_lookup_name_only, &header->name, &elem);
+    if (elem) {
+        *index = (uint64_t)elem->value;
         return AWS_OP_SUCCESS;
     }
 
     /* Check dynamic table */
     aws_hash_table_find(&context->dynamic_table.reverse_lookup, header, &elem);
     if (elem) {
-        const size_t absolute_index = (size_t)elem->value;
+        /* If an element was found, check if it has a value */
+        *found_value = ((const struct aws_http_header *)elem->key)->value.len;
+    } else {
+        /* If not found, check name only table. Don't set found_value, it will be false */
+        aws_hash_table_find(&context->dynamic_table.reverse_lookup_name_only, &header->name, &elem);
+    }
+
+    if (elem) {
+        const uint64_t absolute_index = (uint64_t)elem->value;
         if (absolute_index >= context->dynamic_table.index_0) {
             *index = absolute_index - context->dynamic_table.index_0;
         } else {
             *index = (context->dynamic_table.max_elements - context->dynamic_table.index_0) + absolute_index;
         }
         /* Need to add the static table size to re-base indicies */
-        *index += s_static_header_table_size;
+        *index += STATIC_HEADER_TABLE_SIZE;
         return AWS_OP_SUCCESS;
     }
 
```

Comment:
Github won't let me comment on the `return AWS_OP_ERR;` 1 line down...

Are you sure that an aws_raise_error() occurs in the case that AWS_OP_ERR is returned?

Reply:
No, you make a good point. I'm going to refactor the signature of this function.